replace math round with ceiling for ansible_memtotal_mb + add new variables to display swapfile config

gstorme opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
The Powershell command behind the ansible_memtotal_mb uses a math round of the TotalPhysicalMemory, but it's always rounding the value down.
Because of this, a server with 2GB for example reports 2047MB memotal_mb.
This PR replaces the `round` method with `ceiling` so the actual value is reported by ansible_memtotal_mb.

Fixes #49608 

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
./lib/ansible/modules/windows/setup.ps1

gstorme committed:
replace math round with ceiling to report the actually installed RAM for ansible_memtotal_mb

dagwieers commented:
Hi, thanks for contributing. I think we ought to do the same for the other values.
Can you fix this elsewhere ?

gstorme commented:
I didn't change this for `ansible_swaptotal_mb`, because I noticed this value is always reported as zero on my machines.
Tested on VMWare VM's, OS 2012R2 and 2016, with a custom pagefile set.
So I can change it but I can't test it, won't change anything for me.

I could also change the PR by replacing the WMI class that is used for the `ansible_memtotal_mb` value.
I tried with the property `Capacity` from the `Win32_PhysicalMemory` class (https://docs.microsoft.com/en-us/windows/desktop/cimwin32prov/win32-physicalmemory), and the size reported by this one is correct with a math::round.

Current value (on VMWare):
```
$win32_cs = Get-CimInstance Win32_ComputerSystem
([math]::round($win32_cs.TotalPhysicalMemory / 1024 / 1024))
8191
```

Proposal to change my PR into:
```
$win32_pm = Get-CimInstance Win32_PhysicalMemory
([math]::round($win32_pm.Capacity / 1024 / 1024 ))
8192
```

My tests were done on VMWare & AWS with 2012R2 & 2016.

gstorme commented:
Woops, only just now that I see this comment in the file, so ignore my suggestion to use this class.
`        # Win32_PhysicalMemory is empty on some virtual platforms`

gstorme committed:
add new variable ansible_memtotal to display TotalPhysicalMemory in bytes
add new variable ansible_swap_min to display initial pagefile size
add new variable ansible_swap_max to display maximum pagefile size
removed the variable ansible_swaptotal_mb

gstorme committed:
output ansible_swap_min & ansible_swap_max value in bytes

jborean93 commented on lib/ansible/modules/windows/setup.ps1:
```diff
@@ -295,15 +295,16 @@ if ($gather_subset.Contains("local") -and $factpath -ne $null) {
     $ansible_facts += @{
         # Win32_PhysicalMemory is empty on some virtual platforms
-        ansible_memtotal_mb = ([math]::round($win32_cs.TotalPhysicalMemory / 1024 / 1024))
-        ansible_swaptotal_mb = ([math]::round($win32_os.TotalSwapSpaceSize / 1024 / 1024))
```
We won't be able to just remove a fact without going through the deprecation period of 4 releases so we will need to keep this in.

Looking into `TotalSwapSpaceSize` a bit more, the value itself is defined as part of the open standard [CIM_OperatingSystem](https://docs.microsoft.com/en-us/windows/desktop/CIMWin32Prov/cim-operatingsystem) but I don't believe actually uses it which is why it is null all the time. On Windows 8/10 there is now a file `C:\swapfile.sys` that is used by modern apps but `TotalSwapSpaceSize` doesn't report this.

I think just in case we should keep this value as is and just report null in case this behaviour ever changes in the future. We should also fix this as the value of this property is in kilobytes and not bytes so if it was ever reported then it would have been wrong.

gstorme replied:
I have re-added the ansible_swaptotal_mb fact and changed so it converts KB -> MB

dagwieers replied:
So we cannot remove facts at this time. There is a plan to provide deprecation warnings for deprecated facts (when used). If we have this mechanism we can start thinking about reorganizing some of the facts. (I have a long list of things!)

jborean93 commented on lib/ansible/modules/windows/setup.ps1:
```diff
@@ -295,15 +295,16 @@ if ($gather_subset.Contains("local") -and $factpath -ne $null) {
-        ansible_swaptotal_mb = ([math]::round($win32_os.TotalSwapSpaceSize / 1024 / 1024))
+        ansible_memtotal_mb = ([math]::ceiling($win32_cs.TotalPhysicalMemory / 1024 / 1024))
+        ansible_memtotal = $win32_cs.TotalPhysicalMemory
+        ansible_swap_min = $win32_pf.InitialSize * 1024 * 1024
```
Up to you, if you add the new facts in a separate PR you can backport the rounding fix to stable-2.7. We won't be able to accept new facts on older releases.

gstorme committed:
re-add the ansible_swaptotal_mb fact and fix the conversion of the TotalSwapSpaceSize value from kilobytes to MB

gstorme committed:
indentation fix, replace tab with spaces

jborean93 commented:
Thanks @gstorme for your work on this.

jborean93 commented:
@gstorme unfortunately I've had to revert this commit as it was causing our Windows hosts in Shippable to run a lot slower than normal. I'm unsure why I'm not able to replicate this locally but with these changes affects the performance of 2008 to 2012 R2 on Shippable.

gstorme commented:
Ok, strange.
I tested with these changes on 2012R2 and 2016 with Ansible 2.7, now also ran the test/integration/targets/win_setup/tasks/main.yml tasks on both OS and runs fine locally too.

jborean93 commented:
@gstorme yes I tested it locally and didn’t have any issues with it. No idea why it was an issue with in our CI tests but it definitely made the older hosts slower than before.
